shark_turbine/kernel/wave/codegen.py

```diff
@@ -555,4 +585,74 @@ def handle_sub(emitter: WaveEmitter, node: fx.Node):
 
 @handle_op(reduction)
 def handle_reduction(emitter: WaveEmitter, node: fx.Node):
-    raise NotImplementedError("Reduction: Currently only stub implementation")
+    try:
+        axis, init_args, subgraph, implicit_capture = node.args
+    except ValueError as e:
+        raise ValidationError("Malformed arguments") from e
+
+    # Flatten init_args and get IR values for each of them.
+    flat_init_args, _ = pytree.tree_flatten((init_args))
+    flat_init_args = [cast_py_value(emitter, arg) for arg in flat_init_args]
+
+    # Without scheduling, we assume that we always start at 0.
+    start = arith_d.constant(IndexType.get(), int(0))
+
+    idxc = IndexingContext.current()
+    # For now, we assume that dimensions that have tiling constraints on them,
+    # do not have any other constraints.
+    dim = axis.subs(idxc.subs)
+    end = arith_d.constant(IndexType.get(), int(dim))
+
+    step = None
+    for constraint in emitter.constraints:
+        if isinstance(constraint, TilingConstraint) and constraint.dim == axis:
+            tile_size = constraint.tile_size.subs(idxc.subs)
+            step = arith_d.constant(IndexType.get(), int(tile_size))
+
+    if not step:
+        raise CodegenError(
+            "Could not determine step size for reduction due to missing tiling constraint."
+        )
+
+    forOp = scf_d.ForOp(
+        start,
+        end,
+        step,
+        [a.ir_value for a in flat_init_args],
+    )
+    emitter.induction_vars[axis] = forOp.induction_variable
+    with InsertionPoint(forOp.body):
+        # Add mapping for iter args.
+        subgraph: fx.Graph = emitter.trace.get_subgraph(subgraph)
+        iter_args: list[fx.Node] = get_custom(node).iter_args(subgraph)
+        for i, v in enumerate(forOp.inner_iter_args):
+            emitter.bind_node_proxy(iter_args[i], IRProxyValue(v))
+        captured_vars: list[fx.Node] = get_custom(node).captured_vars(subgraph)
+        for root_v, subgraph_v in zip(implicit_capture, captured_vars):
```

Comment:
QQ: Are both implicit_capture and captured_vars represented as `Placeholder`? Seems like the implicit_captures are the iter_args, would that mean generally implicit_captures == tensors in shared_memory, captured_vars == registers?

Reply:
Sorry, misread the code, I reckon implicit_capture ~= captured_vars. 

But the original question still stand is 
iter_args == tensor for shared memory, and captured_vars == tensor in registers?

Reply:
For the gemm graph, the implicit captures are [a, b] and the captured_vars are [c_reg]. So the implicit captures are global tensors while the captured vars are the iter args of the for loop.